### PR TITLE
Add the ability to throw an error to python debugger

### PR DESCRIPTION
**Pull request by DanielBallaSZTE**
This patch makes it possible to throw an error from the python debugger client using the `throw` command.

JerryScript-DCO-1.0-Signed-off-by: Daniel Balla dballa@inf.u-szeged.hu

### CODE REVIEW

File: jerry-core/vm/vm.c [hunk 1]
```diff
@@ -2539,6 +2539,17 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
                   || JERRY_CONTEXT (debugger_stop_context) == JERRY_CONTEXT (vm_top_context_p)))
           {
             jerry_debugger_breakpoint_hit (JERRY_DEBUGGER_BREAKPOINT_HIT);
+            if (JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_THROW_ERROR_FLAG)
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: I think there are multiple `jerry_debugger_breakpoint_hit`s in the vm. What happns when the others are called and an exception is thrown?

File: jerry-core/vm/vm.c [hunk 2]
```diff
@@ -2539,6 +2539,17 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+            {
+              result = JERRY_CONTEXT (error_value);
+              ecma_value_t to_string_value = ecma_op_to_string (result);
+              ecma_string_t *string_p = ecma_get_string_from_value (to_string_value);
+              ECMA_STRING_TO_UTF8_STRING (string_p, buffer_p, buffer_size);
+              jerry_port_log (JERRY_LOG_LEVEL_ERROR, "Uncaught exception: %.*s\n", buffer_size, buffer_p);
+              ECMA_FINALIZE_UTF8_STRING (buffer_p, buffer_size);
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Do we need to log this? The user sent the request after all.

File: jerry-debugger/jerry-client-ws.py [hunk 3]
```diff
@@ -55,6 +55,8 @@
 JERRY_DEBUGGER_WAIT_FOR_SOURCE = 24
 JERRY_DEBUGGER_OUTPUT_RESULT = 25
 JERRY_DEBUGGER_OUTPUT_RESULT_END = 26
+JERRY_DEBUGGER_STHROW = 27
+JERRY_DEBUGGER_STHROW_END = 28
```
Comment (zherczeg) [inline jerry-debugger/jerry-client-ws.py]: This function sends a byte, not a string. Actually only two values are allowed. What about sending a stop instead?

File: jerry-core/debugger/debugger.h [hunk 4]
```diff
@@ -185,6 +188,8 @@ typedef enum
   JERRY_DEBUGGER_EVAL_PART = 17, /**< next message of evaluating a string */
 
   JERRY_DEBUGGER_MESSAGES_IN_MAX_COUNT, /**< number of different type of input messages */
+  JERRY_DEBUGGER_THROW = 18,
+  JERRY_DEBUGGER_THROW_PART = 19,
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.h]: Where is this symbol is used?

Comment (DanielBallaSZTE) [reply jerry-core/debugger/debugger.h]: I've updated the python client, so these messages are used, when the string is sent.

File: jerry-core/vm/vm.c (hunk 2 above)
Comment (DanielBallaSZTE) [reply jerry-core/vm/vm.c]: I've removed the logging part.

File: jerry-debugger/jerry-client-ws.py (hunk 3 above)
Comment (DanielBallaSZTE) [reply jerry-debugger/jerry-client-ws.py]: I've changed the function to send a 0 or 1 only to the client.

**Comment by DanielBallaSZTE** [comment]
@zherczeg PR Updated, thanks for the review!

File: jerry-core/debugger/debugger.c [hunk 5]
```diff
@@ -272,7 +281,8 @@ jerry_debugger_process_message (uint8_t *recv_buffer_p, /**< pointer the the rec
   if (*expected_message_type_p != 0)
   {
     JERRY_ASSERT (*expected_message_type_p == JERRY_DEBUGGER_EVAL_PART
-                  || *expected_message_type_p == JERRY_DEBUGGER_CLIENT_SOURCE_PART);
+                  || *expected_message_type_p == JERRY_DEBUGGER_CLIENT_SOURCE_PART
+                  || *expected_message_type_p == JERRY_DEBUGGER_THROW_PART);
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.c]: swap the last two lines.

File: jerry-core/debugger/debugger.c [hunk 6]
```diff
@@ -543,7 +558,15 @@ jerry_debugger_process_message (uint8_t *recv_buffer_p, /**< pointer the the rec
...
-      *expected_message_type_p = JERRY_DEBUGGER_EVAL_PART;
+      if (is_eval)
+      {
+        *expected_message_type_p = JERRY_DEBUGGER_EVAL_PART;
+      }
+      else
+      {
+        *expected_message_type_p = JERRY_DEBUGGER_THROW_PART;
+      }
+
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.c]:
```
*expected_message_type_p = (is_eval ? JERRY_DEBUGGER_EVAL_PART
                                    : JERRY_DEBUGGER_THROW_PART);
```

File: jerry-core/debugger/debugger.h [hunk 7]
```diff
@@ -156,6 +157,7 @@ typedef enum
   JERRY_DEBUGGER_OUTPUT_RESULT_END = 26, /**< last output result data */
 
   JERRY_DEBUGGER_MESSAGES_OUT_MAX_COUNT, /**< number of different type of output messages by the debugger */
+  JERRY_DEBUGGER_THROW_STOP = 27, /**< sends the client the info to terminate */
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.h]: comment: "signals the client that the throw is successful" ?

**Comment by DanielBallaSZTE** [comment]
@zherczeg I have updated the PR, thank you for the suggestions!

File: jerry-core/vm/vm.c [hunk 8]
```diff
@@ -2699,9 +2714,19 @@ vm_loop (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
...
+           therefore an evaluation error, or user-created error throw would overwrite it. */
+        ecma_value_t local_err = JERRY_CONTEXT (error_value);
         if (jerry_debugger_send_exception_string ())
         {
           jerry_debugger_breakpoint_hit (JERRY_DEBUGGER_EXCEPTION_HIT);
+          if (JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_THROW_ERROR_FLAG)
+          {
+            result = JERRY_CONTEXT (error_value);
+            ecma_free_value (local_err);
+            goto error;
```
Comment (zherczeg) [inline jerry-core/vm/vm.c]: Would this work here:
```
if (JERRY_CONTEXT (debugger_flags) & JERRY_DEBUGGER_THROW_ERROR_FLAG)
{
  ecma_free_value (local_err);
}
else
{
  JERRY_CONTEXT (error_value) = local_err;
}
```

Comment (zherczeg) [reply jerry-core/vm/vm.c]: Perhaps I would rename `local_err` to `current_error_value`

File: jerry-core/debugger/debugger.c [hunk 9]
```diff
@@ -142,14 +142,15 @@ jerry_debugger_send_backtrace (uint8_t *recv_buffer_p) /**< pointer the the rece
...
  *
  * @return true - if no error is occured
  *         false - otherwise
  */
 static bool
-jerry_debugger_send_eval (const lit_utf8_byte_t *eval_string_p, /**< evaluated string */
-                          size_t eval_string_size) /**< evaluated string size */
+jerry_debugger_send_eval_or_throw (const lit_utf8_byte_t *eval_string_p, /**< evaluated string */
+                                   size_t eval_string_size, /**< evaluated string size */
+                                   bool *is_eval) /**< throw or eval call*/
```
Comment (zherczeg) [inline jerry-core/debugger/debugger.c]: Add `[in,out]` to the comment.

**Comment by DanielBallaSZTE** [comment]
@zherczeg Sorry for the late update, thank you for the review.

**Comment by zherczeg** [comment]
Please rebase. The debugger tests are failing at the moment.

**Comment by DanielBallaSZTE** [comment]
@zherczeg I've rebased the PR .

File: tests/debugger/do_throw.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+function inevitable() {
+  print("An error will be thrown.");
+}
+
+inevitable();
```
Comment (martijnthe) [inline tests/debugger/do_throw.js]: Just curious: why is any code needed in this file for this test case?

Comment (DanielBallaSZTE) [reply tests/debugger/do_throw.js]: Well, we could leave it empty as well.

Comment (zherczeg) [reply tests/debugger/do_throw.js]: I don't think it hurts to keep the current file.

Comment (martijnthe) [reply tests/debugger/do_throw.js]: I was just wondering... I thought perhaps it was needed to get a breakpoint opcode emitted or something.

**Pull request merged**